lib/Sema/CSApply.cpp

```diff
@@ -4196,13 +4196,24 @@ namespace {
       // The declaration we found must be exposed to Objective-C.
       tc.validateDecl(method);
       if (!method->isObjC()) {
-        tc.diagnose(E->getLoc(), diag::expr_selector_not_objc,
-                    foundDecl->getDescriptiveKind(), foundDecl->getFullName())
-          .highlight(subExpr->getSourceRange());
-        tc.diagnose(foundDecl, diag::make_decl_objc,
-                    foundDecl->getDescriptiveKind())
-          .fixItInsert(foundDecl->getAttributeInsertionLoc(false),
-                       "@objc ");
+         // If the method declaration lies in a protocol and we're providing
+         // a default implementation of the method through a protocol extension
+         // and using it as a selector, then bail out as adding @objc to the
+         // protocol might not be the right thing to do and could lead to
+         // problems.
+         if (auto protocolDecl = dyn_cast<ProtocolDecl>(foundDecl->getDeclContext())) {
+             tc.diagnose(E->getLoc(), diag::expr_selector_cannot_be_used,
+                         foundDecl->getBaseName(), protocolDecl->getFullName());
+             return E;
+         }
+          
+         tc.diagnose(E->getLoc(), diag::expr_selector_not_objc,
+                     foundDecl->getDescriptiveKind(), foundDecl->getFullName())
+            .highlight(subExpr->getSourceRange());
+         tc.diagnose(foundDecl, diag::make_decl_objc,
+                     foundDecl->getDescriptiveKind())
+            .fixItInsert(foundDecl->getAttributeInsertionLoc(false),
+                         "@objc ");
```

Comment:
Wrong indentation here too. It's either 4-space (because it's a wrapped statement) or 2-space to align the dots.